target/linux/ramips/mt7621/base-files/lib/upgrade/xiaomi.sh

```diff
@@ -0,0 +1,459 @@
+		log_err "incorrect stock firmware image (rootfs not found)"
+		return 1
+	fi
+
+	XIAOMI_RESTORE_ROOTFS2=true
+	xiaomi_flash_images "$kernel_offset" "$kernel_size" "$rootfs_offset" "$rootfs_size" || {
+		log_err "ERROR: can't revert to stock firmware"
+		return 1
+	}
+	exit 0
+}
+
+platform_do_upgrade_xiaomi() {
+	XIAOMI_FW_FILE=$1
+	local stock_rootfs_size=$2
+	local magic
+	local kernel_mtd
+	local kernel2_mtd
+	local rootfs_mtd
+	local kernel2_part_list
+	local part_name
+
+	XIAOMI_FW_SIZE=$( wc -c "$XIAOMI_FW_FILE" 2> /dev/null | awk '{print $1}' )
+	if [ -z "$XIAOMI_FW_SIZE" ]; then
+		log_err "File '$XIAOMI_FW_FILE' not found!"
+		exit 1
+	fi
+	if [ "$XIAOMI_FW_SIZE" -lt 1000000 ]; then
+		log_err "file '$XIAOMI_FW_FILE' is incorrect"
+		exit 1
+	fi
+
+	kernel_mtd=$( find_mtd_index "$XIAOMI_KERNEL_PART" )
+	if [ -z "$kernel_mtd" ]; then
+		log_err "cannot find mtd partition for '$XIAOMI_KERNEL_PART'"
+		exit 1
+	fi
+	# shellcheck disable=SC2001
+	kernel2_part_list=$( echo "$XIAOMI_KERNEL2_NAMES" | sed 's/|/\n/g' )
+	for part_name in $kernel2_part_list; do
+		kernel2_mtd=$( find_mtd_index "$part_name" )
+		if [ -n "$kernel2_mtd" ]; then
+			XIAOMI_KERNEL2_PART="$part_name"
+			log_msg "Found alt kernel partition '$XIAOMI_KERNEL2_PART'"
+			break
+		fi
+	done
+	rootfs_mtd=$( find_mtd_index "$XIAOMI_ROOTFS_PART" )
+	if [ -z "$rootfs_mtd" ]; then
+		log_err "cannot find mtd partition for '$XIAOMI_ROOTFS_PART'"
+		exit 1
+	fi
+
+	magic=$( get_hexdump_at 0 4 )
+
+	# Flash factory image (uImage header)
+	if [ "$magic" = $MAGIC_UIMAGE ]; then
+		xiaomi_do_factory_upgrade
+		exit $?
+	fi
+
+	# Revert to stock firmware ("HDR1" header)
+	if [ "$magic" = $MAGIC_XIAOMI_HDR1 ]; then
+		if [ -n "$stock_rootfs_size" ]; then
+			XIAOMI_ROOTFS_PARTSIZE=$stock_rootfs_size
+		fi
+		xiaomi_do_revert_stock
+		exit $?
+	fi
+
+	magic=$( get_hexdump_at 0 8 )
+	if [ "$magic" != $MAGIC_SYSUPG ]; then
+		log_err "incorrect image for system upgrading!"
+		exit 1
+	fi
+	log_msg "SysUpgrade start..."
+	local tar_file=$XIAOMI_FW_FILE
+	local board_dir=$( tar tf "$tar_file" | grep -m 1 '^sysupgrade-.*/$' )
+	[ -z "$board_dir" ] && {
+		log_err "board dir not found"
+		exit 1
+	}
+	board_dir=${board_dir%/}
+
+	local control_len=$( (tar xf "$tar_file" "$board_dir/CONTROL" -O | wc -c) 2> /dev/null)
+	if [ "$control_len" -lt 3 ]; then
+		log_err "incorrect stock firmware image (CONTROL not found)"
+		exit 1
+	fi
+	local kernel_len=$( (tar xf "$tar_file" "$board_dir/kernel" -O | wc -c) 2> /dev/null)
+	if [ "$kernel_len" -lt 1000000 ]; then
+		log_err "incorrect stock firmware image (kernel not found)"
+		exit 1
+	fi
+	local rootfs_len=$( (tar xf "$tar_file" "$board_dir/root" -O | wc -c) 2> /dev/null)
+	if [ "$rootfs_len" -lt 1000000 ]; then
+		log_err "incorrect stock firmware image (rootfs not found)"
+		exit 1
+	fi
+
+	if [ -n "$XIAOMI_KERNEL2_PART" ]; then
+		tar Oxf "$tar_file" "$board_dir/kernel" | mtd -f write - "$XIAOMI_KERNEL2_PART"
+		if [ $? -ne 0 ]; then
```

Comment:
```suggestion
		if ! tar Oxf "$tar_file" "$board_dir/kernel" | mtd -f write - "$XIAOMI_KERNEL2_PART"; then
```
?

Reply:
the final idea is to have every script passed with shellscript so that will probably be changed in the future